src/util/address.rs

```diff
@@ -406,6 +407,140 @@ impl Payload {
             } => script::Script::new_witness_program(version, prog)
         }
     }
+
+    /// Creates a pay to (compressed) public key hash payload from a public key
+    #[inline]
+    pub fn p2pkh(pk: &ecdsa::PublicKey) -> Payload {
+        Payload::PubkeyHash(PubkeyHash::hash(&pk.serialize()))
+    }
+
+    /// Creates a pay to script hash P2SH payload from a script
+    #[inline]
+    pub fn p2sh(script: &script::Script) -> Result<Payload, Error> {
+        if script.len() > MAX_SCRIPT_ELEMENT_SIZE {
+            return Err(Error::ExcessiveScriptSize);
+        }
+        Ok(Payload::ScriptHash(ScriptHash::hash(&script[..])))
+    }
+
+    /// Create a witness pay to public key payload from a public key
+    pub fn p2wpkh(pk: &ecdsa::PublicKey) -> Result<Payload, Error> {
+        if !pk.compressed {
```

Comment:
Could this be represented in type?

Reply:
I couldn't work out a way, did you have a concrete idea?

Reply:
I do: pls see my commend below

Reply:
I meant taking `struct CompressedPublicKey(secp256k1::PublicKey)` which guarantees the key being compressed. We could then push checking to the boundary and also have a helpful `compress(self) -> CompressedPublicKey` method. (Might make sense to implement in `secp256k1`.)